slidechain/slidechain_test.go

```diff
@@ -265,209 +265,222 @@ func TestImport(t *testing.T) {
 }
 
 func TestEndToEnd(t *testing.T) {
-	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
 	defer cancel()
-	withTestServer(ctx, t, func(ctx context.Context, db *sql.DB, s *submitter, sv *httptest.Server, ch *protocol.Chain) {
-		hclient := &horizon.Client{
-			URL:  "https://horizon-testnet.stellar.org",
-			HTTP: new(http.Client),
-		}
-		root, err := hclient.Root()
-		if err != nil {
-			t.Fatalf("error getting horizon client root: %s", err)
-		}
-		accountID, seed, err := custodianAccount(ctx, db, hclient)
-		if err != nil {
-			t.Fatalf("error creating custodian account: %s", err)
-		}
-		c := &Custodian{
-			seed:          seed,
-			AccountID:     *accountID,
-			S:             s,
-			DB:            db,
-			hclient:       hclient,
-			InitBlockHash: ch.InitialBlockHash,
-			imports:       sync.NewCond(new(sync.Mutex)),
-			exports:       sync.NewCond(new(sync.Mutex)),
-			network:       root.NetworkPassphrase,
-			privkey:       custodianPrv,
-		}
-		c.launch(ctx)
-
-		// Prepare Stellar account to peg-in funds and txvm account to receive funds.
-		amount := 5 * xlm.Lumen
-		exporterPub, exporterPrv, err := ed25519.GenerateKey(nil)
-		if err != nil {
-			t.Fatalf("error generating txvm recipient keypair: %s", err)
-		}
-		var exporterSeed [32]byte
-		copy(exporterSeed[:], exporterPrv)
-		exporter, err := keypair.FromRawSeed(exporterSeed)
-		err = stellar.FundAccount(exporter.Address())
-		if err != nil {
-			t.Fatalf("error funding account %s: %s", exporter.Address(), err)
-		}
+	// TODO(debnil): Test non-native assets.
+	var tests = []struct {
+		inputAmount  xlm.Amount
+		exportAmount xlm.Amount
+	}{
+		{5 * xlm.Lumen, 5 * xlm.Lumen},
+		{5 * xlm.Lumen, 3 * xlm.Lumen},
+	}
+	for _, tt := range tests {
+		withTestServer(ctx, t, func(ctx context.Context, db *sql.DB, s *submitter, sv *httptest.Server, ch *protocol.Chain) {
+			hclient := &horizon.Client{
+				URL:  "https://horizon-testnet.stellar.org",
+				HTTP: new(http.Client),
+			}
+			root, err := hclient.Root()
+			if err != nil {
+				t.Fatalf("error getting horizon client root: %s", err)
+			}
+			accountID, seed, err := custodianAccount(ctx, db, hclient)
+			if err != nil {
+				t.Fatalf("error creating custodian account: %s", err)
+			}
+			c := &Custodian{
```

Comment:
Could we factor out the custodian launching + exporter setup so we only do that once, and use the same custodian / exporter for all the test cases?